drivers/power_domain/power_domain_intel_adsp.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pm/device.h>
+#include <pm/device_runtime.h>
+#include <pm/intel/power_gating_intel_adsp.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(power_domain_intel_adsp, LOG_LEVEL_INF);
+
+static int poll_for_register_delay(uint32_t reg, uint32_t mask, uint32_t val, uint64_t us)
+{
+	if (!WAIT_FOR((sys_read32(reg) & mask) != val, us, k_busy_wait(1))) {
+		LOG_ERR("%s poll timeout reg %u mask %u val %u us %u", __func__, reg, mask, val,
+			(uint32_t)us);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int pg_set_device_power_enable(struct pg_registers reg, bool power_enable, uint64_t us)
+{
+	int ret;
+	uint32_t spa_mask = BIT(reg.SPA_bit);
+	uint32_t cpa_mask = BIT(reg.CPA_bit);
+
+	if (power_enable)
+		sys_write32(sys_read32(reg.wr_address) | spa_mask, reg.wr_address);
+	else
+		sys_write32(sys_read32(reg.wr_address) & ~(spa_mask), reg.wr_address);
+
+	/* Check if powered on. */
+	ret = poll_for_register_delay(reg.rd_address, cpa_mask, spa_mask, us);
+
+	if (ret)
+		LOG_WRN("%s warning: timeout", __func__);
+
+	return ret;
+}
+
+#ifdef CONFIG_PM_DEVICE_POWER_DOMAIN
```

Comment:
`POWER_DOMAIN_DEVICE`s are created under this ifdef. There are no power domains instantiated when this option is disabled. Most of the power managed devices are still implemented in SOF, which means that the power domain could turn off if none of the already zephyr-integrated ones are in use, while still some of the sof-implemented ones are. I believe that this verbose config helps with the possible confusion there. If it's a no-go it can be omitted.